lang/c/src/schema.h

```diff
@@ -76,12 +78,30 @@ struct avro_link_schema_t {
 	avro_schema_t to;
 };
 
+struct avro_bytes_schema_t {
+  struct avro_obj_t obj;
+  avro_logical_schema_t *logical_type;
+};
```

Comment:
I think to provide "first class" logical schema support, this approach would need to be reversed: instead of primitive/complex schemas containing logical schemas, the logical schemas should wrap their underlying types in some sense

by first class support, I mean that [avro_type_t](https://github.com/apache/avro/blob/78540a1e59d8556226b9ff1484ff17212137de62/lang/c/src/avro/basics.h#L29-L47) doesn't have any logical schema branches...which means that [avro_value_iface::get_type](https://github.com/apache/avro/blob/78540a1e59d8556226b9ff1484ff17212137de62/lang/c/src/avro/value.h#L100-L103) alone wouldn't be enough to determine if a schema is logical

Reply:
Thanks for looking at the PR. 
AFAIU, logical type is a non-mandatory annotation of a physical type, which is why [avro_type_t] is left without logical schema branches. For this reason, [avro_logical_schema_t] is implemented as a standalone decorator. Doesn't this make sense?

Reply:
I had started looking into ipc and logical support shortly before running into this PR, and managed to open #2891 as an alternative where I add decimal support by creating a new `AVRO_DECIMAL` `avro_type_t` branch

in that PR, the decimal support is added to the "value api" instead of the legacy "datum api", but in an optional way: users can call `get_decimal` on the new decimal values, or continue calling `get_bytes` or `get_fixed`

fwiw, the approach you've taken here is more similar to existing implementations from what I've seen: i.e. most implementations have a `getType()` and then a `getLogicalType()` method that users are expected to call if they're interested in logical types

in my opinion though (feel free to disagree), the distinction between logical and non-logical types is only useful for the spec and wire-format: from a user's pov, they're probably more interested in working with timestamps and decimals instead thinking about the underlying int and "big-endian encoded two's-complement" representations